src/util/futurepool.rs

```diff
@@ -0,0 +1,356 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// This mod implemented a wrapped future pool that supports `on_tick()` which is driven by
+/// tasks and is invoked no less than the specific interval.
+
+use std::fmt;
+use std::cell::{Cell, RefCell, RefMut};
+use std::sync::{mpsc, Arc};
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::thread;
+use std::time::Duration;
+use futures::Future;
+use futures_cpupool::{self as cpupool, CpuFuture, CpuPool};
+
+use util;
+use util::time::Instant;
+use util::collections::HashMap;
+
+pub trait Context: fmt::Debug + Send {
+    /// Will be invoked periodically (no less than specified interval).
+    /// When there is no task, it will NOT be invoked.
+    fn on_tick(&mut self) {}
+}
+
+/// A delegator to wrap `Context` to provide `on_tick` feature.
+#[derive(Debug)]
+struct ContextDelegator<T: Context> {
+    tick_interval: Duration,
+    inner: RefCell<T>,
+    last_tick: Cell<Option<Instant>>,
+}
+
+impl<T: Context> ContextDelegator<T> {
+    fn new(context: T, tick_interval: Duration) -> ContextDelegator<T> {
+        ContextDelegator {
+            tick_interval,
+            inner: RefCell::new(context),
+            last_tick: Cell::new(None),
+        }
+    }
+
+    fn context_mut(&self) -> RefMut<T> {
+        self.inner.borrow_mut()
+    }
+
+    fn on_task_finish(&self) {
+        let now = Instant::now_coarse();
+        let last_tick = self.last_tick.get();
+        if last_tick.is_none() {
+            // set last_tick when the first future is resolved
+            self.last_tick.set(Some(now));
+            return;
```

Comment:
What if the second task comes long after the first?  I think we'd better call `on_tick` when the first future is resolved according to the definition of `last_tick`.

Reply:
I think the scenario you suggest is rare. In real world we have more than one tasks come together at first, i.e. TiDB loading schemas.